Implement rm(..., force=true)

eschnett opened the pull request:
Implement a `force=true` option for `rm` that does not report an error when a path does not exist.
When returning an error, ensure it is always a `SystemError` and not a `UVError`.
Add test cases.
Update documentation.

tkelman commented on base/docs/helpdb.jl:
```diff
@@ -8590,9 +8590,9 @@ As a special case, if `x` is an `AbstractString` (for textual MIME types) or a `
 
 doc"""
-    rm(path::AbstractString; recursive=false)
+    rm(path::AbstractString; force=false, recursive=false)
```
rst docs need updating for this too. make signature consistent then run genstdlib.jl

tkelman commented on base/file.jl:
```diff
@@ -88,19 +88,26 @@ end
+            if force && isa(err, UVError) && err.prefix=="unlink" && err.code==Base.UV_ENOENT
+                return
+            end
+            systemerror(:unlink, true, extrainfo=path)
```
what does the normal uverror from rm look like here? should we be rethrowing the error code too?

eschnett replied:
The UVError says "unlink" and has `errno` as code. I'll change things to preserve this in the `SystemError`.

tkelman commented:
tests are having issues on windows, but I like it. sort of unfortunate that `force` is the most consistent name here with coreutils but that isn't really describing what this option does all that accurately.

eschnett commented:
Updated rst file.

tkelman commented on base/file.jl:
```diff
@@ -88,19 +88,30 @@ end
+                # convert UVError to SystemError
+                rethrow(SystemError(err.prefix, err.code, path))
+            end
+            rethrow()
```
I guess this should be really rare so in most cases we get the useful extrainfo

eschnett replied:
I don't know whether it can actually occur, so I wanted to leave the exception alone.

tkelman commented on base/file.jl:
```diff
@@ -88,19 +88,30 @@ end
+                    return
+                end
+                # convert UVError to SystemError
+                rethrow(SystemError(err.prefix, err.code, path))
```
should this maybe be `throw` instead of `rethrow`?

eschnett replied:
I don't know. The actual `unlink` system call occurred elsewhere, hence the `rethrow`. I'd be happy either way.

tkelman commented on base/file.jl:
```diff
@@ -88,19 +88,30 @@ end
+                    return
+                end
+                # convert UVError to SystemError
+                rethrow(SystemError(err.prefix, err.code, path))
```
darn, the `UVError` codes are different from the `SystemError` codes:
```
julia> rm("nonexistent")
ERROR: SystemError (with nonexistent): unlink: Unknown error -2
 [inlined code] from stream.jl:1024
 in unlink at file.jl:370
 in __rm#52__ at file.jl:95
 in eval at ./boot.jl:263
```

eschnett replied:
They seem to be negatives of each other?

eschnett replied:
Anyway, other functions (e.g. `chmod`) also return `UVError` objects. Probably best to leave this alone in this patch.

tkelman replied:
most are negatives of the errno.h versions, but only on unix: https://github.com/JuliaLang/libuv/blame/julia-uv0.11.26/include/uv-errno.h

my `errno_h.jl` on Windows looks quite different: https://gist.github.com/b3e80b58153607dfb9a9

kind of a shame to lose the extrainfo of the path in the error, but we can leave that to a later PR to figure out

eschnett replied:
On Thu, Dec 3, 2015 at 11:44 PM, Tony Kelman <notifications@github.com>
wrote:

> In base/file.jl
> <https://github.com/JuliaLang/julia/pull/14248#discussion_r46648880>:
>
> >      if islink(path) || !isdir(path)
> > -        @windows_only if (filemode(path) & 0o222) == 0; chmod(path, 0o777); end # is writable on windows actually means "is deletable"
> > -        unlink(path)
> > +        try
> > +            @windows_only if (filemode(path) & 0o222) == 0; chmod(path, 0o777); end # is writable on windows actually means "is deletable"
> > +            unlink(path)
> > +        catch err
> > +            if isa(err, UVError)
> > +                if force && err.code==Base.UV_ENOENT
> > +                    return
> > +                end
> > +                # convert UVError to SystemError
> > +                rethrow(SystemError(err.prefix, err.code, path))
>
> most are negatives of the errno.h versions, but only on unix:
> https://github.com/JuliaLang/libuv/blame/julia-uv0.11.26/include/uv-errno.h
>
> my errno_h.jl on Windows looks quite different:
> https://gist.github.com/b3e80b58153607dfb9a9
>
> kind of a shame to lose the extrainfo of the path in the error, but we can
> leave that to a later PR to figure out
>
I guess the easiest thing to do later on is to create a `SystemError` from
scratch, which would re-derive the error code from the current `errno`.

-erik

-- 
Erik Schnetter <schnetter@gmail.com>
http://www.perimeterinstitute.ca/personal/eschnetter/

eschnett committed:
Implement rm(..., force=true)

Implement a `force=true` option for `rm` that does not report an error when a path does not exist.
Add test cases.
Update documentation.

stevengj commented:
Now that this is merged, presumably #14145 should be updated?

eschnett commented:
It should.
